### ISSUE TITLE
Make RAND_DRBG a public OpenSSL API

### ISSUE THREAD

**Issue by mspncp** [openssl/openssl#4403]
This is a followup to #4328 and #4402. As I said previously on many occasions, it is my vision and my belief that the RAND_DRBG API should become a public object oriented OpenSSL API.

The RAND_DRBG API is clearly more low-level than the RAND_METHOD API, so it makes sense to place the RAND_DRBG declarations in a separate header `<openssl/rand_drbg.h>` distinct from `<openssl/rand.h>` and make it clear in the documentation that this API is intended for “back end” developers, like somebody who tries to integrate a hardware RNG as entropy source and not for the standard user.

This issue does not compete with issue #4394. Having a simplified API for plugging in standard or user-defined entropy sources is a good way for 90% of the use cases. But there are also cases, where one needs access to the full power of the RAND_DRBG API. For example, in one of our products the default RAND_METHOD is mapped to a DRBG which reseeds on every generate request (prediction_resistance = true) and pulls its entropy from a smartcard  (/dev/random is only used as additional input). Such special cases can only be handled with the full API.

The RAND_DRBG should be a full blown "class", i.e. the structure should be opaque with the necessary accessors. To be truly object oriented, the rwlock needs to be owned by the object, and the locking managed entirely by the RAND_DRBG's "methods" (please excuse my C++-speak) and not by the caller.

The libcrypto library is a low level general purpose crypto library. So it should provide a toolbox of useful algorithms and not make any assumptions about the way it will be used by the programmer. In particular, it should enable but not force the programmer to use the legacy RAND_METHOD api.

**Comment by mspncp** [comment]
## Roadmap

Now that PR #4328 has finally been merged, here is my private roadmap of the most notable things to be done, before the RAND_DRBG interface can be published as official OpenSSL API. The list of coworkers mentioned below is ad hoc and comes from my wishful thinking. Anybody else willing to work on this subject may speek up, he's welcome.

### Add master DRBG

This is already work in progress, see PR #4402. I will rebase and freshen it up shortly, and continue working on it. This will be my principal occupation for the near future.

The code is already functional, but it would probably be a good idea to add unit tests to verify that the seed chaining works. I will try whether I can implement the tests myself or whether I need help struggling with the test framework. Or maybe one of the core members volunteers to implement the tests for the seed chaining?

### Improve Multiple entropy source handling

@paulidale and I had some private discussions about this, and he created Issue #4394 for this. Some vague ideas exist, but it's not the "egg of Columbus" yet. Any suggestions and good ideas are welcome.

### Finetune the RAND_POOL API  (and make it opaque)

In PR #4328 it turned out that the RAND_POOL API is not perfect yet and might require changes. This could be coordinated with the work on Issue #4394 - Improve Multiple entropy source handling. I volunteer to work on this, with input from @paulidale, @kroeckx e.a.

### Implement locking inside the RAND_DRBG

Currently, ony the global DRBGs have a lock, and locking is managed by the callers, i.e., the methods from the RAND_METHOD api. @kaduk and I discussed this privately and we agreed that in the future the locking should be managed by the RAND_DRBG class itself, but it should be an optional feature because it is not needed by the SSL CHILD DRBGs. Maybe Ben would like to undertake this task and create a PR for this? I will be happy to assist.

### Make the RAND_DRBG 'class' opaque

When making the locking 'opaque', Ben could also make the rest of the structure opaque and check for accessor completeness.

### Publish the RAND_DRBG and RAND_POOL interfaces

Although I want them published, I consider the RAND_DRBG and RAND_POOL interfaces as 'backend' interfaces, which are not intended for the normal OpenSSL user. For that reason I would not place the API into `<openssl/rand.h>`, but create a new header file, say  `<openssl/rand_drbg.h>`.

* Move the RAND_DRBG and RAND_POOL interfaces from  `<internal/rand.h>` to `<openssl/rand_drbg.h>`
* Add complete documentation. This is a lot, but much can be borrowed and adopted from the FIPS 2.0 user guide and the NIST standard.

**Comment by kroeckx** [comment]
Other items I would like to add:
- Sort out when to use the public vs the private DRBG. We might need to change some internal APIs so they know which DRBG to use.
- Do we want want a public, private or both DRBG per ssl object? I guess there are various tradeoffs.

**Comment by mspncp** [comment]
In some sense, the ssl DRBG is already private. So I‘m not sure whether having two SSL specific DRBGs wouldn‘t be a little bit overengineered. But one could think about chaining the ssl DRBGs to the private DRBG instead of the public.

**Comment by kroeckx** [comment]
My understanding of private vs public is that the generated data will become private or public. While you seem to be thinking about private in C++ terms as part of the class.

**Comment by mspncp** [comment]
No, not in the C++ sense. My understanding was simply, that the private DRBG is used for private keys and the public DRBG for everything else. But I did not ask Rich personally, who introduced it.

**Comment by mspncp** [comment]
To be honest, I didn’t give it a deeper thought yet. So I agree with you that the DRBG usage (public vs private) needs to be sorted out.

**Comment by richsalz** [comment]
Yes, it was a separate DRBG for generating private keys.  Not in the C++ sense.

**Comment by kroeckx** [comment]
So I've split part of that out into https://github.com/openssl/openssl/issues/4641

**Comment by mspncp** [comment]
It's all in my head already, but I had little time to pursue this task recently. Within the next days I hope to update this thread with an outline of what remains to be done. I'll do my best to get my first "personal feature" finished before the feature freeze. :-)

**Comment by mspncp** [comment]
> The RAND_DRBG should be a full blown "class", i.e. the structure should be opaque with the necessary accessors. To be truly object oriented, the rwlock needs to be owned by the object, ~~and the locking managed entirely by the RAND_DRBG's "methods" (please excuse my C++-speak) and not by the caller~~.

I changed my mind about the last part of this statement. Having the locking managed entirely inside the RAND_DRBG class made the implementation too clumsy, and the gain was little, since it is better to encourage unlocked per-thread instances in the first place. In order to modify the shared DRBGs it is necessary to have a locking api, which I added in pull request #5294

**Comment by mspncp** [comment]
## To lock or not to lock

Ben Kaduk braught up the question in https://github.com/openssl/openssl/pull/5294#issuecomment-364468261, whether it desirable to have an official API for locking RAND_DRBG instances, like I proposed:

```
   int RAND_DRBG_lock(RAND_DRBG *drbg)
   int RAND_DRBG_unlock(RAND_DRBG *drbg)
   int RAND_DRBG_enable_locking(RAND_DRBG *drbg)
```

This creates somehow a precedent case, since it exposes internal locking structures in a public API and should be discussed by the @openssl/committers.

Indeed one can argue with equal justification for one direction or the other. The current implementation in `master` appears somewhat undecided and goes in neither direction. Let me try to explain what I mean and describe the possible variants.

### Locking variants

The original FIPS DRBG implementation did not contain any locking code, the locking was entirely handled by the RAND method wrapping the FIPS DRBG singleton. The `drbg->lock` member was added only recently (in 12fb8c3d2dd) to the RAND_DRBG structure, at the very beginning when the FIPS DRBG code was ported to master.

Initially, it seemed very natural to me that every DRBG should have its own lock and that locking is enabled only if necessary, which currently means only for the three shared DRBG instances. I envisioned an object oriented RAND_DRBG class, so I argued that the RAND_DRBG instances should handle locking completely internally, if their locking is activated.

#### Make the RAND_DRBG instances intrinsically thread safe

I even started implementing this, but soon realized that this approach had some disadvantages. The thread safety is achieved at the price of practically every method having to take the `drbg->lock` internally. This not only bloats the code (even more since locking is optional), but also has the disadvantage, that the lock can't be held among multiple API calls. Another complication is that currently some public methods call others, for example `RAND_DRBG_bytes()` calls `RAND_DRBG_generate()`. To prevent double locking, the latter would have to be split into a private non-locking implementation `rand_drbg_generate` and a public locking wrapper `RAND_DRBG_generate()`.

The overall gain is little, since except for the three global DRBG's locking is disabled anyway because it is preferrable to have per-thread instances. This alltogether convinced me, that this is not the right approach.

#### Provide an (optional) external locking mechanism

This is what #5294 tries to achieve: An optional locking mechanism which has an official API and can be used by consumers of the RAND_DRBG class. Since locking is needed and implemented for the three shared instances, I see no reason why it should not be available as a reusable public API. Of course, SSL/TLS is the main focus, but after all, `libcrypto` is a _general purpose_ cryptography toolkit.


#### Provide an undocunmented locking mechanism explicitely for the RAND api

This is the status quo in master: The RAND_DRBG contains a `drbg->lock` member, but it is enabled and managed entirely by the RAND callbacks. They do it by reaching directly into the members of RAND_DRBG, without accessors and information hiding.

Also, locking is enabled explicitely for the three shared DRBG's and this is hardcoded. (When locking was introduced, there was only a single instance.) 

An important Question to ask here is:

> Why does every RAND_DRBG instance need a lock, if this lock is only used for three hardcoded instances, and in a very global way? Why not remove the `drbg->lock` members entirely and have a single global lock protecting those three instances?


### Conclusion

In my opinion both directions, to lock or not to lock, have their pros and cons. Either choice is ok to me, but one should make a decision and then  do it right. So there are two possibilities:

- Variant 1: Drop locking support entirely from RAND_DRBG
- Variant 2: Provide a consistent official API to access the option locking mechanism

Variant 2 is attempted in pull request #5294. If locking is useful for the RAND api, why shouldn't is also be useful and hence reusable in other situations. If it works, why not let the user use it? It's a general purpose crypto library.

Variant 1 means to remove the `drbg->lock` member entirely from the `RAND_DRBG` structure. As a replacement, protect access to the global DRBG's via `RAND_bytes()` e.a. by a single global lock. That's how it was in the FIPS code.

### Public access to the shared DRBGs

A related question is: should the application be allowed to access the shared DRBG's, i.e. should `RAND_DRBG_get0_{master,public,private}()` be part of the public API and what is the use case.

Well, the use case is to meet very specific requirements for the random generator. For example, enforce prediction resistance, replace the entropy source by a different one, etc. It may not be a standard use case for SSL/TLS, but surely one for a general purpose crypto library. And this believe comes not from my imagination, it is taken from my practice.

If the user can access the shared DRBG's then we can either provide a locking mechanism, or we have to add the requirement in the documentation that changing properties of the shared DRBGs like reseed intervals must be done before creating additional threads.

**Comment by kroeckx** [comment]
Do any other DRBGs other than the 3 global ones ever need locking? Do we want to support applications that want to create their own DRBG used by multiple threads? If so, should we do the locking, or should the application do it?

**Comment by mspncp** [comment]
This is exactly my question. And it sounds like you are in favor of variant 1.

**Comment by kaduk** [comment]
Variant 1 seems to be most consistent with the historical/traditional OpenSSL mentality, to the extent that I am qualified to judge.  But I don't know that we need to be entirely bound by history.

**Comment by richsalz** [comment]
I don't think history should guide us here.  *We* need to lock the objects, and providing a lock to users will make this thread-safety-possible.

**Comment by t-j-h** [comment]
We need to provide thread safety for everything we declare is thread-safe to access. 
For example, we don't currently support multi-threaded access to a single SSL, but SSL_CTX is meant to be thread-safe (only not all functions are).

We certainly don't want to add locking calls to every function in the library.

**Comment by vdukhovni** [comment]
The RNG needs to be thread-safe.

**Comment by kroeckx** [comment]
I think what @t-j-h and @vdukhovni are saying is the "Make the RAND_DRBG instances intrinsically thread safe", maybe we can call that variant 0.

**Comment by vdukhovni** [comment]
As much as possible, avoid global state, the RNG could use thread-local storage, for example.

**Comment by vdukhovni** [comment]
Or always make syscalls for entropy, which is thread-safe.

**Comment by kroeckx** [comment]
You also need to update the internal state and you don't want different threads to be changing your internal state.

**Comment by mspncp** [comment]
It is no question that the three shared DRBGs need locking. Variant 1 suggests that it makes more sense to have a global lock than to have a per-instance lock which is allowed to be used only for three hardcoded instances.

**Comment by mspncp** [comment]
Variant 0 is pretty heavyweight. In particular if one wants to make _all_ objects intrinsically thread safe. There will be a lot of code duplication in the absence of class inheritance. I believe OpenSSL should facilitate thread safety wherever possible, but not implement it by all means.

**Comment by mspncp** [comment]
> We need to provide thread safety for everything we declare is thread-safe to access.

One possible solution would be to declare the standard RAND api as thread safe (which it already is) and the low level RAND_DRBG api as not thread safe (i.e., variant 1). 

The documentation will make very clear, that the RAND_DRBG api is a low level api which is not intended for every day  usage ("Just use RAND_bytes() and don't worry about any details."). If the developer  (e.g, myself) for some reason does need low level access to the three shared DRBGs used by the default OpenSSL RAND method (for modifying the default behaviour), then he can do it safely during initialization before creating threads.

**Comment by mspncp** [comment]
Rethinking #5294, I noticed that my suggestion to replace the per-drbg locks by a global one for variant 1 was premature: It works fine for RAND_bytes(), but makes things more complicated when reseeding chained DRBGs, like the SSL DRBGs. So the current implementation in #5294 is ok. The real question is not actually wether the locking api should be implemented. It is useful, since using it is more expressive than reaching into the rand_drbg_st struct and using the lower level thread api.

What remains is the question: Now that we have it it and it works, should we make it available to the user, or should it remain for internal use in libcrypto and libssl only. And that's really a strategical decision only, which I leave up to you:

- Variant 1: Use the RAND_DRBG locking api internally in libcrypto and libssl only
- Variant 2: Add the RAND_DRBG locking api to the official API header (<openssl/rand_drbg.h>, t.b.d.)

**Comment by richsalz** [comment]
My view is that if we make this API public, then we should make it thread-safe from the beginning.  Yes, this means almost everything will eventually have to have locking methods, just like almost everything as upref/release methods.

**Comment by mspncp** [comment]
One of the advantages of having multiple DRBG instances is performance and scalability, since per-thread instances can be operated without locking. So wouldn't it be better to encourage per-thread usage of a lightweight RAND_DRBG api instead of providing a 'fat' locking implementation?

**Comment by mspncp** [comment]
People who don't care about threading and performance may just stick with RAND_bytes().

**Comment by mspncp** [comment]
Thank you everybody for your detailed feedback. There was a great diversity of opinions, but it was precisely your diversity that gave me a lot of ideas and helped me overcome my own unsureness and figure out what I think now is the optimal solution. Let me restate the variants:

- Variant 1: Use the RAND_DRBG locking api internally in libcrypto and libssl only.
- Variant 2: Add the RAND_DRBG locking api to the official API header.
- Variant 3: Make RAND_DRBG fully thread-safe unsing internal locking. (*)

> (*) Variant 3 was originally introduced as variant 0 by Kurt, but considering the combined implementation and documentation effort, the number 3 is more appropriate. Variant 0 would be "do nothing and leave everything as it is."

First of all, it became clear that there is some reluctance to add the locking functions to the public api, because this would creat a precedence case. These concerns are justified, so I decided to drop variant 2.

Second, it is common agreement that the global RNGs should be thread-safe. But they already are thread-safe, if accessed through the RAND api. So variant 3 is imho a bit overengineered. The RAND_DRBG api is a lower level building block and in my opinion does not have to be thread-safe, just as all the other low level crypto primitives aren't thread-safe either. In view of @t-j-h's remark,

> We need to provide thread-safety _for everything we declare is thread-safe to access_.

it is justified to declare the RAND api as thread-safe, whereas the RAND_DRBG api is not. 

Consequently, I suggest to go for variant 1. This means, finish pull request #5294 but leave the locking api for internal use only. The three global DRBGs can be accessed in a thread-safe manner via the RAND api, all other DRBG instances are intentionally not thread-safe.  Configuring the global DRBGs (reseed interval etc.) is safe only during application initialization before creating threads.

If at any time the need arises to switch to full thread-safety for RAND_DRBG (which I doubt will happen), this can be done without changing the public api (because the locking functions are not part of it).

**Comment by kaduk** [comment]
On "first" (re)read Variant 1 seems to make sense.
Especially since many of our public structures (e.g., ```SSL *```) are only safe to use from a single thread already absent application-level locking.
So, if an application did want to use a single RAND_DRBG from multiple threads (e.g., as a parent for seeding), it would need to either embed the RAND_DRBG into a parent structure with a lock, or stick a lock in EX_DATA or something like that, and do the appropriate locking around its accesses, including a purpoted get_entropy_from_parent() rewrite.

**Comment by mspncp** [comment]
If this would be a common use case, then one could still publish the locking functions. Because rewriting `get_entropy_from_parent()` is a hack. But I doubt this will ever happen.

Apropos hack: the user can just add prototypes for the locking functions, because they are listed in libcrypto.num anyway.

**Comment by kroeckx** [comment]
Why are they in libcrypto.num? I don't see them being used outside the drbg code.

**Comment by mspncp** [comment]
Sorry, I provided outdated information: The symbols of the locking api are _not_ listed in libcrypto.num anymore. I finally removed them before merging #5294.

**Comment by mspncp** [comment]
Well, I _did_ remove the locking api from libcrypto.num, but they were added back in by mkdef.pl. This is going to be fixed in #5375 .
